File: python/2022/23/solution.py
import common
from collections import defaultdict

lines = common.read_file().splitlines()
elves = set()

for y in range(len(lines)):
    for x in range(len(lines[0])):
        if lines[y][x] == '#':
            elves.add((x, y))

def choose_move(i, elves, x, y):
    # all neighbors
    ns = [
        (x-1, y-1), (x, y-1), (x+1, y-1),
        (x+1, y), (x+1, y+1),
        (x, y+1), (x-1, y+1),
        (x-1, y)]

    # if no elves close, don't move
    for n in ns:
        if n in elves:
            break
    else:
        return None

    # neighbors in order for north, south, west, east
    dirs = [
        ns[:3],
        ns[4:7],
        ns[6:] + ns[:1],
        ns[2:5],
    ]

    # positions after move in order for north, south, west, east
    dir_results = [
        (x, y-1),
        (x, y+1),
        (x-1, y),
        (x+1, y),
    ]

    # check moves in all directions, order based on iteration
    for di in range(len(dirs)):
        idx = (i+di) % len(dirs)
        if all([p not in elves for p in dirs[idx]]):
            return dir_results[idx]

    return None

curr_elves = elves
i = 0
while True:
    moved = False
    moves = dict()
    destinations = defaultdict(lambda: 0)
    for e in curr_elves:
        move = choose_move(i, curr_elves, *e)
        if move is not None:
            moves[e] = move
            destinations[move] += 1
    new_elves = set()
    for e in curr_elves:
        if e not in moves:
            new_elves.add(e)
            continue
        m = moves[e]
        if destinations[m] > 1:
            new_elves.add(e)
            continue
        new_elves.add(m)
        moved = True
    curr_elves = new_elves
    i += 1

    # part 1
    if i == 10:
        min_x = min([e[0] for e in curr_elves])
        max_x = max([e[0] for e in curr_elves])
        min_y = min([e[1] for e in curr_elves])
        max_y = max([e[1] for e in curr_elves])

        print((max_x-min_x+1)*(max_y-min_y+1)-len(elves))

    if not moved:
        break

# part 2
print(i)

File: go/2022/23/solution.go
package main

import (
	"aoc/utils"
	"aoc/utils/set"
)

type Pos struct {
	x, y int
}

func main() {
	lines := utils.ReadLines()
	elves := set.Set[Pos]{}
	for y, line := range lines {
		for x, c := range []rune(line) {
			if c == '#' {
				elves.Add(Pos{x, y})
			}
		}
	}

	currElves := elves
	i := 0
	for {
		moved := false
		moves := map[Pos]Pos{}
		destinations := map[Pos]int{}
		for e := range currElves {
			move := chooseMove(i, currElves, e.x, e.y)
			if move != nil {
				moves[e] = *move
				destinations[*move] += 1
			}
		}

		newElves := set.Set[Pos]{}
		for e := range currElves {
			m, ok := moves[e]
			if !ok || destinations[m] > 1 {
				newElves.Add(e)
			} else {
				newElves.Add(m)
				moved = true
			}
		}

		currElves = newElves
		i += 1

		// part 1
		if i == 10 {
			e := currElves.Peek()
			minX, maxX, minY, maxY := e.x, e.x, e.y, e.y
			for p := range currElves {
				if p.x < minX {
					minX = p.x
				}
				if p.x > maxX {
					maxX = p.x
				}
				if p.y < minY {
					minY = p.y
				}
				if p.y > maxY {
					maxY = p.y
				}
			}

			println((maxX-minX+1)*(maxY-minY+1) - len(elves))
		}

		if !moved {
			break
		}
	}

	// part 2
	println(i)
}

func chooseMove(i int, elves set.Set[Pos], x int, y int) *Pos {
	// TODO
	neighbors := []Pos{
		{x - 1, y - 1},
		{x, y - 1},
		{x + 1, y - 1},
		{x + 1, y},
		{x + 1, y + 1},
		{x, y + 1},
		{x - 1, y + 1},
		{x - 1, y},
	}

	found := false
	for _, n := range neighbors {
		if elves.Contains(n) {
			found = true
			break
		}
	}

	if !found {
		return nil
	}

	dirs := [][]Pos{
		neighbors[:3],
		neighbors[4:7],
		append(append([]Pos{}, neighbors[6:]...), neighbors[:1]...),
		neighbors[2:5],
	}

	dir_results := []Pos{
		{x, y - 1},
		{x, y + 1},
		{x - 1, y},
		{x + 1, y},
	}

	for di := range dirs {
		idx := (i + di) % len(dirs)
		all := true
		for _, p := range dirs[idx] {
			if elves.Contains(p) {
				all = false
				break
			}
		}

		if all {
			return &dir_results[idx]
		}
	}

	return nil
}
